### PR TITLE
Fix | The SqlConnection.FireInfoMessageEventOnUserErrors when set to true throws an exception

### LINKED ISSUE #2388: 5.2.0: using sqlConnection.FireInfoMessageEventOnUserErrors = true generates "Object reference not set to an instance of an object." exception
### Describe the bug
When using sqlConnection.FireInfoMessageEventOnUserErrors = true, a "print" sql command always throws an exception

**Exception message:**
Object reference not set to an instance of an object.

**Stack trace:**
   at Microsoft.Data.SqlClient.SqlCommand.GetCurrentBatchCommand()
   at Microsoft.Data.SqlClient.TdsParser.FireInfoMessageEvent(SqlConnection connection, SqlCommand command, TdsParserStateObject stateObj, SqlError error)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean isAsync, Int32 timeout, Boolean asyncWrite)
   at Microsoft.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String methodName)
   at Microsoft.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at SqlClientBug.Program.Main(String[] args) in E:\NAS2\COBEMAR\2022\dotNet\SqlClientBug\Program.cs:line 31

### To reproduce
Include a complete code listing (or project/solution) that we can run to reproduce the issue.

Partial code listings, or multiple fragments of code, will slow down our response or cause us to push the issue back to you to provide code to reproduce the issue.

```c#
namespace SqlClientBug
{
   internal class Program
    {
        static void Main(string[] args)
        {
            SqlConnectionStringBuilder sb = new();

            sb.TrustServerCertificate = true;
            sb.DataSource = ".";
            sb.InitialCatalog = "tempdb";
            sb.IntegratedSecurity = true;
            sb.ApplicationName = "SqlClientBug";
            
            using SqlConnection sqlConnection = new(sb.ConnectionString);
            sqlConnection.FireInfoMessageEventOnUserErrors = true;
            sqlConnection.Open();
            using SqlCommand cmd = sqlConnection.CreateCommand();
            cmd.CommandType = System.Data.CommandType.Text;
            cmd.CommandText = "print 'OK';";
            try
            {
                cmd.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine(ex.Message);
                Console.Error.WriteLine("--------------------");
                Console.Error.WriteLine(ex.StackTrace);
            }
        }
    }
}
```

### Expected behavior
As in previous versions: no Exception

### Further technical details
Microsoft.Data.SqlClient version: 5.2.0
.NET target: .NET 8.0
SQL Server version: SQL Server 2022
Operating system: Windows 11

**Additional context**
No problem when using Microsoft.Data.SqlClient version 5.1.5

### PR DESCRIPTION
**Pull request by arellegue**
Fixes #2388

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/SqlCommand/SqlCommandFireInfoMessageEventOnUserErrorsTest.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+
+namespace Microsoft.Data.SqlClient.ManualTesting.Tests
+{
+    public class SqlCommandFireInfoMessageEventOnUserErrorsTest
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/SqlCommand/SqlCommandFireInfoMessageEventOnUserErrorsTest.cs]: FireInfoMessageEventOnUserErrorsShouldSucceed test now included in ConnectivityTest class.

**Comment by JRahnama** [comment]
The proposed fix seems satisfactory for merging as it stands. However, it's important to note that this portion of the code was altered by #1825. Given that it introduces a new feature to the driver, it's essential to conduct a thorough review to ensure that the implementation behaves as intended in this particular scenario.

cc: @Wraith2

**Review by David-Engel — APPROVED** [review]

**Comment by DavoudEshtehari** [comment]
Fix the title with a proper description, please.

**Comment by Wraith2** [comment]
I'm intrigued. How do we managed to have to reach `GetCurrentBatchCommand` without executing an rpc? It should have gone through `GetRPCObject` in order to send anything to the server and so the rpc should not non-null.

**Comment by JRahnama** [comment]
> I'm intrigued. How have we managed to have to reach `GetCurrentBatchCommand` without executing an rpc? It should have gone through `GetRPCObject` in order to send anything to the server and so the rpc should be non-null.

I am curious to understand why `GetCurrentBatchCommand` was added to that particular exception? GetRPCObject has not been called at that stage. I tested with Batch and it also came back null. I was wondering if we should let the default value of null to be passed on that call?

**Comment by Wraith2** [comment]
If there is never a way to call the caller of `GetCurrentBatchCommand` in this situation and expect there to be a batch command then we should not make the call to `GetCurrentBatchCommand` and just use null. If it's possible then this nullable approach is fine. It seems odd that we could raise an exception through the command reporting path without even calling the server.

**Review by DavoudEshtehari — COMMENTED** [review]
LGTM, waiting for the resolution by @JRahnama and @Wraith2.

**Review by JRahnama — APPROVED** [review]

**Comment by Wraith2** [comment]
Ok. I debugged through it and understand the situation now. The library only needs to use an RPC object internally if there are parameters and in this case there aren't so it follows a path that simply sends the text without first building any sort of RPC. It'd be good to get this sort of breakdown as part of the fix so those of us following along at home can understand why the fix works.

It does raise a question or what happens if you try to setup a BatchCommand with a set of commands that have no parameters, would errors be reported correctly or crash?

**Review by Wraith2 — APPROVED** [review]

**Pull request merged**